File: common/src/main/java/dev/mayaqq/estrogen/client/cosmetics/ui/CosmeticsClaimScreen.java
package dev.mayaqq.estrogen.client.cosmetics.ui;

import com.simibubi.create.foundation.gui.Theme;
import dev.mayaqq.estrogen.client.cosmetics.service.CosmeticsApi;
import net.minecraft.client.Minecraft;
import net.minecraft.client.gui.GuiGraphics;
import net.minecraft.client.gui.components.EditBox;
import net.minecraft.client.gui.screens.Screen;
import net.minecraft.network.chat.CommonComponents;
import net.minecraft.network.chat.Component;
import net.minecraft.util.FormattedCharSequence;

public class CosmeticsClaimScreen extends BaseCosmeticsScreen {

    private Component info;

    public CosmeticsClaimScreen(Screen parent) {
        super(parent);
    }

    @Override
    protected void contentInit(int x, int y, int width, int height) {
        this.claimButton.asDisabled();

        int buttonX = x + width / 2 - 50;
        int buttonY = (int) (y + height * 0.75f - 10);

        // Claim Code Input
        addRenderableOnly(UIComponents.box(buttonX, buttonY, 100, 16));
        var textBox = addRenderableWidget(new EditBox(font, buttonX + 2, buttonY + 3, 96, 14, CommonComponents.EMPTY));
        textBox.setBordered(false);

        // Claim Button
        addRenderableWidget(UIComponents.textButton(CosmeticUI.CLAIM_BUTTON)
                .withPosition(buttonX, buttonY + 30)
                .withBounds(100, 20)
                .withCallback(() -> claim(textBox.getValue()))
        );
    }

    @Override
    protected void renderWindow(GuiGraphics graphics, int mouseX, int mouseY, float partialTicks) {
        super.renderWindow(graphics, mouseX, mouseY, partialTicks);

        int x = (int) (this.width * 0.25f) + PADDING * 2;
        int y = PADDING * 3;
        int width = this.width - x - PADDING;

        int infoX = x + width / 2;
        int infoY = y;

        for (FormattedCharSequence text : font.split(CosmeticUI.CLAIM_DESCRIPTION, width)) {
            graphics.drawCenteredString(font, text, infoX, infoY, Theme.i(Theme.Key.BUTTON_IDLE));
            infoY += 10;
        }

        if (this.info != null) {
            infoY += 10;
            graphics.drawCenteredString(font, this.info, infoX, infoY, Theme.i(Theme.Key.BUTTON_FAIL));
        }
    }

    public void claim(String code) {
        CosmeticsApi.claimReward(code)
                .thenAcceptAsync(status -> {
                    if (status == CosmeticsApi.StatusCode.OK) {
                        Minecraft.getInstance().tell(() -> {
                            Screen screen = Minecraft.getInstance().screen;
                            if (screen != null) {
                                screen.onClose();
                            }
                        });
                    } else {
                        this.info = CosmeticUI.getClaimMessage(status);
                    }
                });
    }

    public static void open() {
        Minecraft.getInstance().setScreen(new CosmeticsClaimScreen(Minecraft.getInstance().screen));
    }
}

File: src/common/main/kotlin/dev/mayaqq/estrogen/client/content/screen/cosmetics/CosmeticsClaimScreen.kt
package dev.mayaqq.estrogen.client.content.screen.cosmetics

import dev.mayaqq.cynosure.helpers.McClient
import dev.mayaqq.cynosure.helpers.McFont
import dev.mayaqq.cynosure.text.CommonText
import dev.mayaqq.cynosure.text.TextUtils.splitLines
import dev.mayaqq.cynosure.text.unaryMinus
import dev.mayaqq.estrogen.client.content.screen.EstrogenButton
import dev.mayaqq.estrogen.client.content.screen.EstrogenMenuScreen
import dev.mayaqq.estrogen.client.cosmetics.CosmeticAPI
import dev.mayaqq.estrogen.client.cosmetics.StatusCode
import net.minecraft.client.Minecraft
import net.minecraft.client.gui.GuiGraphics
import net.minecraft.client.gui.components.EditBox
import net.minecraft.client.gui.screens.Screen
import net.minecraft.network.chat.Component


class CosmeticsClaimScreen(previous: Screen?) : CosmeticsBaseScreen(previous, -"gui.estrogen.cosmetics.claim") {

    var info: Component? = null

    val bClaim = EstrogenButton.Builder(EstrogenButton.TextRenderer(-"gui.estrogen.cosmetics.claim.button")) {
        CosmeticAPI.claimReward(codeBox.value).thenAcceptAsync { status ->
                if (status === StatusCode.OK) {
                    Minecraft.getInstance().tell {
                        McClient.screen?.onClose()
                    }
                } else {
                    this.info = getClaimMessage(status)
                }
            }
    }.color(EstrogenMenuScreen.transWhite)

    val codeBox = EditBox(McFont, 0, 0, 0, 21, CommonText.EMPTY)

    override fun baseInit() {
        super.baseInit()
        val fakeThirdWidth = (width - 30) / 3
        val claimButton = bClaim.bounds(
            20 + fakeThirdWidth + 80,
            height - 35,
            fakeThirdWidth * 2 - 160,
            25
        ).disabled(codeBox.value.isEmpty()).buildAndAdd()

        codeBox.x = (20 + fakeThirdWidth + 80) + 2
        codeBox.y = height - 70 + 2
        codeBox.width = (fakeThirdWidth * 2 - 160) - 4
        codeBox.setResponder {
            claimButton.disabled = it.isEmpty()
        }
        codeBox.add()
    }

    override fun beforeRender(graphics: GuiGraphics, mouseX: Int, mouseY: Int, partialTick: Float) {
        val fakeThirdWidth = (width - 30) / 3
        val x = (fakeThirdWidth * 2) + 20
        var y = ((height - 90) / 2) - (McFont.lineHeight * 3)

        (-"gui.estrogen.cosmetics.claim.description").splitLines().forEach { line ->
            graphics.drawCenteredString(McFont, line, x, y, 0xFFFFFF)
            y += McFont.lineHeight
        }

        info?.let {
            y += McFont.lineHeight
            graphics.drawCenteredString(font, it, x, y, 0xFFFFFF)
        }
    }
}